src/cudadev/plugin-SiPixelClusterizer/SiPixelROCsStatusAndMappingWrapperESProducer.cc

```diff
@@ -30,8 +30,10 @@ void SiPixelROCsStatusAndMappingWrapperESProducer::produce(edm::EventSetup& even
   {
     std::ifstream in(data_ / "cablingMap.bin", std::ios::binary);
     in.exceptions(std::ifstream::badbit | std::ifstream::failbit | std::ifstream::eofbit);
-    SiPixelROCsStatusAndMapping obj;
-    in.read(reinterpret_cast<char*>(&obj), sizeof(SiPixelROCsStatusAndMapping));
+    // We use default alignment
+    auto objBuffer = std::make_unique<std::byte[]>(SiPixelROCsStatusAndMapping::computeDataSize(pixelgpudetails::MAX_SIZE));
+    SiPixelROCsStatusAndMapping obj(objBuffer.get(), pixelgpudetails::MAX_SIZE);
```

Comment:
On a further thought, in the memory ownership model we've mainly discussing about there is a distinction between an owner object and a non-owning "view". I realize the SoA here looks a lot like such non-owning view, in which case leaving the memory management outside of the SoA is completely understandable. But I wonder if it would be possible to craft a generic owner class in this case?

Reply:
I think this is an important place to look. Currently, while running `cudadev` with those parameters: `./cudadev --maxEvents 10000 --numberOfThreads 16 --numberOfStreams 24`, we can see in NSight that CPU threads spend most of their time waiting for the mutex of the caching memory allocator. The GPU and copy engine are not saturated. 

A single memory handler per producer allocating a single chuck of memory and cutting it into one or multiple views should reduce the number of allocations by a factor of a few units (up to 8 here: https://github.com/cms-patatrack/pixeltrack-standalone/blob/master/src/cudadev/CUDADataFormats/SiPixelDigisCUDA.cc#L8-L15 ). I am planning to look into this now as there is a good performance gain opportunity here.

Eventually, this memory manager could also handle the various copies in host, pinned, host and devices memory, and the transfers.

Reply:
> Currently, while running `cudadev` with those parameters: `./cudadev --maxEvents 10000 --numberOfThreads 16 --numberOfStreams 24`, we can see in NSight that CPU threads spend most of their time waiting for the mutex of the caching memory allocator.

Is the bottleneck really the mutex in the caching allocator? The last time I looked into that (which admittedly was nearly 2 years ago and in CMSSW) the CPU-side bottleneck was the mutex in the CUDA itself. Things have certainly evolved since then, but I'd like to see the evidence.

What GPU are you testing on? In my tests on V100 `cudadev` saturates around `--numberOfThreads 10`, so at least there going much beyond will just result in more lock contention. Luckily(?) this setup also reaches 92-98 % GPU utilization as reported by `nvidia-smi` (whatever that really means).

Reply:
I had a look with NSight. Here is a snapshot:

![image](https://user-images.githubusercontent.com/37585813/131964228-8e8f3d7e-b8d9-45be-bcbf-fec4d705dbca.png)

This is with a T4 (with 2 Intel(R) Xeon(R) Gold 6338 for a total of 64 cores/128 threads). NSight saw 23 threads in total, so CPU is not used at the maximum. The very bottom blue/purple bars show the usage of the GPU processing and copy engine, respectively. We are clearly not saturating the GPU (~50%, judging by eye).

It's easy to differentiate the CUDA locking from the memory manager locking: CUDA uses a pthread_rwlock_wrlock and NSight shows the CUDA call under the system call (see the red `cudaMem...` on the 4th thread on the left), while the memory manager locking uses pthread_mutex_lock. 

All can be checked with a tooltip stack trace:

![image](https://user-images.githubusercontent.com/37585813/131965569-1dca83c9-606d-4e9c-ad8f-bc893d86495c.png)

I tried to minimize the locking in the memory manager by deferring potentially expensive calls our of the critical section, but this did not help. One of such expensive calls can be visualized here:
![image](https://user-images.githubusercontent.com/37585813/131965962-56c330da-0b03-4f45-be74-9d93df64760d.png)

We can see all other threads waiting for the mutex in this snapshot.

I added deferred calls here: https://github.com/ericcano/pixeltrack-standalone/commit/40abde015f3b51aebb71b6296724c3e294cd8519 that's a risk free change, but it did not help much.

I also moved the cudaEventRecord out of the critical section, but this requires to change to a 2 pass strategy, which actually reduces performance (probably adding to the contention as each pass requires a lock): https://github.com/ericcano/pixeltrack-standalone/commit/a3fdba0e873bced08e3e8bd37d9bbca6de1ee8de

So, in all, reducing the number of allocations by gathering SoA columns in a single block should improve the performance (looking into this now).

Reply:
This is very interesting. Have you tried what happens with smaller number of CPU threads, say between 5 and 10? I'd expect T4 to saturate with at most the same amount of CPU threads as V100, so I'd expect 23 to suffer heavily from lock contention.

Nevertheless I agree it is clear that the caching allocator must be improved. Given also the discussion in #216, I opened an issue to discuss that further https://github.com/cms-patatrack/pixeltrack-standalone/issues/218.

Reply:
I moved the relevant code out to another branch: https://github.com/ericcano/pixeltrack-standalone/tree/memoryManagement